### ISSUE TITLE
Add option to use fixed port for communication 

### ISSUE THREAD

**Issue by mayurikini** [altugbakan/rs-tftpd#35]
Currently create_multi_socket function binds to a random port. 
I propose adding an additional parameter to allow the use of a fixed port. This will help ensure predictable communication by always binding the socket to the same local port when needed.

**Comment by altugbakan** [comment]
There is already a `—single-port` flag, does it not do what you want?

**Comment by mayurikini** [comment]
Single port also uses a random port, can we add option to use fixed assigned port?

**Comment by stappersg** [comment]
On Fri, Dec 13, 2024 at 10:32:42PM -0800, Mayuri Kini wrote:
> Currently create_multi_socket function binds to a random port. 

Yes, random source port at client side is a good thing.
In fact it is an essential part of TFTP.


> I propose adding an additional parameter to allow the use of a fixed port.

Make parameter optional and non default.
Prevent that TFTP servers need to keep track of client IP address,
allow them to just keep track of client port number.


> This will help ensure predictable communication by always binding the
> socket to the same local port when needed.

Please provide some examples of "when needed".



Groeten
Geert Stappers


P.S.

Yes, patches welcome.
-- 
Silence is hard to parse

**Comment by stappersg** [comment]
On Fri, Dec 13, 2024 at 11:22:01PM -0800, Mayuri Kini wrote:
>  ... can we add option to use fixed assigned port? 

Why?

**Comment by naikaakash** [comment]
Hello @stappersg,
The concern is the use of random port when transmitting data after the initial request.

Initial request sent to server on port 69. (server is listening on it)

Then the server responds to client with DATA using a “random” port. This random port is then used in the session to transmit more data. So now server has to be listening on this newly created port(currently set to single port) aswell.
This operation of listening on new port is fine and is working as expected. But the randomness brings a challenge from OS level. The firewall needs to open this port(or a range of ports) for this communication.

We need to find a way to support a way to specify a port (or a range) to which the upcoming message after the initial request can be served rather than relying on a random port address.

**Comment by naikaakash** [comment]
I tested some things locally. With -s flag set. It will not choose a random port. It would use 69 or whatever port was specified in -p.

Please confirm if this is expected behaviour. If so, then I think Mayuri should be good with her request without a change.

**Comment by altugbakan** [comment]
@naikaakash Yes that is intended, hence my question about the request.

**Comment by stappersg** [comment]
On Tue, Dec 17, 2024 at 10:20:16AM -0800, Mayuri Kini wrote:
> No change needed. Closing this issue.

Thanks
